### PR TITLE
Fix remainder type promotion

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48668 Fix remainder type promotion**
* #48278 Fix `fmod` type promotion

Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
In the case where the second argument is a python number, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
### This PR:
In the case where the second argument is a python number, the dtype of result is determined by type promotion of both inputs.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```

## BC-breaking tests:
[Nightly](https://www.internalfb.com/intern/testinfra/testconsole/testrun/6192449517872347/)
[This PR](https://www.internalfb.com/intern/testinfra/testconsole/testrun/3659174740128654/)
This PR doesn't introduce any new failure or timeout model.

Differential Revision: [D25869136](https://our.internmc.facebook.com/intern/diff/D25869136)

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -3803,13 +3803,19 @@ def forward(self, input, other):
         y = torch.randn(1, 2, 1)
         self.run_test(RemainderModel(), (x, y))
 
+    # Note: Integral Tensor remainder to floating-point scalar (not Tensor)
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3806]: this is bc-breaking change. Can we still make this case working for the original test? 

Otherwise, existing pipeline might just break. cc: @spandantiwari, @BowenBao

Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3806]: @ejguan could you try removing the casting at here

https://github.com/pytorch/pytorch/blob/dee82ef3ea1ed2cae4e891f071adff03cb4a6e0d/torch/onnx/symbolic_opset9.py#L2598

and see if it works for the original test case? The was workaround of onnx export for remainder since it was not adhering to type promotion before.

File: test/onnx/expect/TestOperators.test_remainder.expect [hunk 2]
```diff
@@ -3,41 +3,30 @@ producer_name: "pytorch"
...
     output: "2"
-    name: "Cast_0"
-    op_type: "Cast"
-    attribute {
-      name: "to"
-      i: 1
-      type: INT
-    }
+    name: "Div_0"
+    op_type: "Div"
```
Comment (ejguan) [inline test/onnx/expect/TestOperators.test_remainder.expect:10]: Hi @BowenBao 
I updated the expected IR of remainder, since the operator of `remainder` is changed. And the `test/onnx/test_operator.py  TestOperators.test_remainder` is passed now.

**Comment by ejguan** [comment]
@mruberry 
Could you review this PR and #48278 , since the ONNX errors on both PRs have been mitigated.

File: test/test_binary_ufuncs.py [hunk 3]
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
...
-        # out
-        out = torch.empty(0, device=device, dtype=dtype)
-        torch.fmod(x, zero, out=out)
-        self.assertEqual(out.size(), torch.Size([10, 10]))
-        self.assertTrue(torch.all(out.isnan()))
-        # in-place
-        x.fmod_(zero)
-        self.assertTrue(torch.all(x.isnan()))
+    def test_fmod_remainder_by_zero_float(self, device, dtype):
+        fn_list = [(torch.fmod, lambda x, mod: x.fmod_(mod)),
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1363]: This enumeration of functions is cool but I think it can be simplified to:

`(torch.Tensor.fmod_, torch.Tensor.remainder_)`

Comment (mruberry) [reply test/test_binary_ufuncs.py:1363]: Also, in general we should tuples for iterables. Lists are OK when the iterable has to be mutable.

File: test/test_binary_ufuncs.py [hunk 4]
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
...
+                   (torch.remainder, lambda x, mod: x.remainder_(mod))]
+        for fns in fn_list:
+            # check floating-point tensor fmod/remainder to zero is nan on both CPU and GPU
+            x = make_tensor((10, 10), device=device, dtype=dtype, low=-9, high=9)
+            zero = torch.zeros_like(x)
+
+            self.assertTrue(torch.all(fns[0](x, 0.0).isnan()))
+            self.assertTrue(torch.all(fns[0](x, zero).isnan()))
+            # out
+            out = torch.empty(0, device=device, dtype=dtype)
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1373]: I don't think you need to test the out or inplace variants for this behavior.

File: test/test_binary_ufuncs.py (hunk 3 above)
Comment (mruberry) [reply test/test_binary_ufuncs.py:1363]: Note that the behavior of returning NaN when the denominator of fmod or remainder is zero is consistent with the C++ standard but inconsistent with NumPy. That's OK. We should prefer the C++ or Python standards where applicable.

File: test/test_binary_ufuncs.py [hunk 5]
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
...
-                torch.fmod(x, zero, out=out)
-                self.assertEqual(out == 4294967295, x >= 0)
-                self.assertEqual(out == -1, x < 0)
-                self.assertEqual(out.size(), torch.Size([10, 10]))
-                # in-place
-                x_.fmod_(zero)
-                self.assertEqual(x_ == 4294967295, x >= 0)
-                self.assertEqual(x_ == -1, x < 0)
+    def test_fmod_remainder_by_zero_integral(self, device, dtype):
+        fn_list = [(torch.fmod, lambda x, mod: x.fmod_(mod)),
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1385]: See comments above.

File: test/test_binary_ufuncs.py [hunk 6]
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
...
+                   (torch.remainder, lambda x, mod: x.remainder_(mod))]
+        for fns in fn_list:
+            # check integral tensor fmod/remainder to zero
+            x = make_tensor((10, 10), device=device, dtype=dtype, low=-9, high=9)
+            zero = torch.zeros_like(x)
+            # out
+            out = torch.empty(0, device=device, dtype=dtype)
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1392]: Once again for special cases it shouldn't be necessary to test variant behavior like out=.

File: test/test_binary_ufuncs.py [hunk 7]
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
...
+            # In-place
+            x_ = x.clone()
+            # RuntimeError on CPU
+            if device == 'cpu':
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1396]: `if self.device_type == 'cpu'` or `if torch.device(device).type == 'cpu'`

File: test/test_binary_ufuncs.py [hunk 8]
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
...
-                self.assertTrue(torch.all(x.fmod(zero) == value))
-                # out
-                torch.fmod(x, zero, out=out)
-                self.assertTrue(torch.all(out == value))
-                self.assertEqual(out.size(), torch.Size([10, 10]))
-                # in-place
-                x_.fmod_(zero)
-                self.assertTrue(torch.all(x_ == value))
+                if dtype == torch.int64:
+                    self.assertEqual(fns[0](x, zero) == 4294967295, x >= 0)
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1406]: Would you elaborate on our integer CUDA behavior? That is, where are these numbers coming from?

File: test/test_binary_ufuncs.py [hunk 9]
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
...
+                    self.assertEqual(out.size(), torch.Size([10, 10]))
+                    # in-place
+                    fns[1](x_, zero)
+                    self.assertTrue(torch.all(x_ == value))
 
     @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=False, include_bool=False, include_complex=False))
-    def test_fmod(self, device, dtype):
+    def test_fmod_remainder(self, device, dtype):
         # Use numpy as reference
         def _reference_implementation(x, mod):
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1431]: Once again, "reference_implementation" seems like a misnomer because this function actually performs the test.

File: test/test_binary_ufuncs.py [hunk 10]
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
...
+                    np_mod = mod.cpu().numpy()
+                else:
+                    np_mod = mod
+                exp = fns[2](np_x, np_mod)
+                exp = torch.from_numpy(exp)
 
-            res = torch.fmod(x, mod)
-            common_dtype = torch.promote_types(res.dtype, exp.dtype)
-            exp = exp.to(common_dtype)
+                res = fns[0](x, mod)
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1444]: Analogous comments about type handling here as in the fmod PR

Comment (mruberry) [reply test/test_binary_ufuncs.py:1444]: It would be more readable if the tuple was unpacked as `fn, inplace, ref` immediately instead of referencing its parts with indices.

File: test/test_binary_ufuncs.py [hunk 11]
```diff
@@ -1471,44 +1480,6 @@ def _reference_implementation(x, mod):
             _reference_implementation(x, mod)
             _reference_implementation(x_nc, mod)
 
-    @onlyCPU
-    @dtypes(torch.float, torch.long)
-    def test_remainder(self, device, dtype):
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1476]: Is it really OK to just remove this test and the following one? The new tests do not seem to cover everything that it does.

Comment (ejguan) [reply test/test_binary_ufuncs.py:1476]: I think the following tests cover:
1. dtype for float and long
2. Tensor remainder to scalar or scalar tensor (positive and negative)
3. Tensor remainder to another tensor (broadcasting)
The only thing left, which I will put into the new test cases, is negative divisor.

File: test/test_binary_ufuncs.py (hunk 7 above)
Comment (mruberry) [reply test/test_binary_ufuncs.py:1381]: Since so much of PyTorch's code is copy-pasted it's important to use the blessed patterns, even when another, possibly simpler pattern, will work. For example, if a test tried to do `if device == 'cuda':` that would fail, because device could be `cuda:0` or `cuda:1` when the test is run on a CUDA device. That's why it's important to check the device's type using `self.device_type` or constructing a device with `torch.device(device)` and then acquiring its type.

File: test/test_binary_ufuncs.py [hunk 12]
```diff
@@ -1359,72 +1359,79 @@ def test_rdiv(self, device, dtype):
...
+        fn_list = (torch.fmod, torch.remainder)
+        for fn in fn_list:
+            # check integral tensor fmod/remainder to zero
+            x = make_tensor((10, 10), device=device, dtype=dtype, low=-9, high=9)
+            zero = torch.zeros_like(x)
+            # RuntimeError on CPU
+            if device == 'cpu':
+                with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
+                    fn(x, zero)
+            # Different value for different dtype on CUDA:
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1384]: This note is very helpful. In a future PR we should probably update the documentation for fmod and remainder to include this, too.

File: test/test_binary_ufuncs.py [hunk 13]
```diff
@@ -1359,72 +1359,79 @@ def test_rdiv(self, device, dtype):
...
+            fns_list = ((torch.fmod, torch.Tensor.fmod_, np.fmod),
+                        (torch.remainder, torch.Tensor.remainder_, np.remainder))
+            for fn, inplace_fn, ref in fns_list:
+                np_x = x.cpu().numpy()
+                np_mod = mod.cpu().numpy() if torch.is_tensor(mod) else mod
+                exp = ref(np_x, np_mod)
+                exp = torch.from_numpy(exp)
+
+                res = fn(x, mod)
+                # Different Type promotion between numpy and torch for
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1411]: To enforce consistent integer->float type promotion you can wrap the NumPy references with

https://github.com/pytorch/pytorch/blob/573f4aa3529371e9d30a0884d56eac0e6d3ddd71/torch/testing/_internal/common_methods_invocations.py#L284

Then you wouldn't have to deal with this

**Comment by mruberry** [comment]
Hey @ejguan, this PR looks pretty good. I made a few inline comments which should be easy to deal with.

One last thing to do on this PR is to write a "BC-breaking Note:" above the PR summary explaining what the previous behavior was vs. what the new behavior is, and to label this PR as "BC-breaking." These notes help the release notes writers for each version of PyTorch. This BC-break is very minor, but we might as well do it for completeness. 

I reviewed our operator database for uses of fmod and remainder that might actually change the behavior of existing models and didn't find any, so we're probably OK. If you're curious about a more thorough BC-breaking change test, however, let me know and I can walk through the process.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 14]
```diff
@@ -116,8 +116,8 @@ void div_kernel(TensorIterator& iter) {
 }
 
 void remainder_kernel(TensorIterator& iter) {
-  if (isIntegralType(iter.dtype(), /*includeBool*/ false)) {
-    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "remainder_cpu", [&]() {
+  if (isIntegralType(iter.common_dtype(), /*includeBool*/ false)) {
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:119]: What happens if remainder is called with a bool argument?

File: test/test_binary_ufuncs.py (hunk 7 above)
Comment (ejguan) [reply test/test_binary_ufuncs.py:1381]: It makes a lot of sense!

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 15]
```diff
@@ -116,8 +116,8 @@ void div_kernel(TensorIterator& iter) {
 }
 
 void remainder_kernel(TensorIterator& iter) {
-  if (isIntegralType(iter.dtype(), /*includeBool*/ false)) {
-    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "remainder_cpu", [&]() {
+  if (isIntegralType(iter.common_dtype(), /*includeBool*/ true)) {
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:119]: Like with fmod, let's exclude bool for remainder, too. NumPy also promotes bool inputs to remainder.

File: aten/src/ATen/native/cuda/BinaryRemainderKernel.cu [hunk 16]
```diff
@@ -12,8 +12,8 @@
 namespace at { namespace native {
 
 void remainder_kernel_cuda(TensorIterator& iter) {
-  if (isIntegralType(iter.dtype(), /*includeBool*/ false)) {
-    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "remainder_cuda", [&]() {
+  if (isIntegralType(iter.common_dtype(), /*includeBool*/ true)) {
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:15]: Remove bool here, too.

File: test/test_binary_ufuncs.py [hunk 17]
```diff
@@ -1354,63 +1354,70 @@ def test_rdiv(self, device, dtype):
...
-            except RuntimeError as e:
-                self.assertRegex(str(e), "result type (Half|Float|Double|Long) "
-                                         "can't be cast to the desired output "
-                                         "type (Bool|Byte|Char|Short|Int|Long)")
+            fns_list = ((torch.fmod, torch.Tensor.fmod_, np.fmod),
+                        (torch.remainder, torch.Tensor.remainder_, np.remainder))
+            for fn, inplace_fn, ref_fn in fns_list:
+                np_x = x.cpu().numpy()
+                np_mod = mod.cpu().numpy() if torch.is_tensor(mod) else mod
+                ref_fn = np_binary_ufunc_type_promotion_wrapper(ref_fn)
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1402]: See fmod PR for a proposal simplifying this.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 18]
```diff
@@ -128,7 +128,7 @@ void remainder_kernel(TensorIterator& iter) {
       });
     });
   } else {
-    AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "remainder_cpu", [&]() {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.common_dtype(), "remainder_cpu", [&]() {
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:131]: Move this dispatch out of the else branch.

File: aten/src/ATen/native/cuda/BinaryRemainderKernel.cu [hunk 19]
```diff
@@ -23,7 +23,7 @@ void remainder_kernel_cuda(TensorIterator& iter) {
       });
     });
   } else {
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:25]: Move this dispatch out of the else branch.

File: torch/_torch_docs.py [hunk 20]
```diff
@@ -7306,17 +7306,17 @@ def merge_dicts(*dicts):
...
-When :attr:`other` is a tensor, the shapes of :attr:`input` and
-:attr:`other` must be :ref:`broadcastable <broadcasting-semantics>`.
+Supports :ref:`broadcasting to a common shape <broadcasting-semantics>`,
+:ref:`type promotion <type-promotion-doc>`, and integer and float inputs.
 
-Note:
+.. note::
     Complex inputs are not supported. In some cases, it is not mathematically
     possible to satisfy the definition of a modulo operation with complex numbers.
+    Check the note from :func:`torch.fmod` for zero divisor.
```
Comment (mruberry) [inline torch/_torch_docs.py:7315]: "See :func:\`torch.fmod\` for how division by zero is handled."

**Review by mruberry — APPROVED** [review]
Nice work, @ejguan. Like with fmod, I've inlined a few simple changes to the kernels and a small docs update. Also like with fmod we should run an additional BC-breaking test on this diff before landing it.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 18 above)
Comment (ejguan) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:131]: I think we still need to keep it within the else block. Otherwise, we will dispatch two times for this operation.
